00_find_bad_subs.py

```diff
@@ -0,0 +1,70 @@
+"""finds bad subjects and saves them as json.
+
+based on either sensitvity index d'
+or binomial test against chance.
+the latter works bc of a clever task design
+To reproduce, you have to delete the bad subs from config.
+"""
+# %%
+# Imports
+import pandas as pd
+from scipy.stats import binom_test
+
+from config import FPATH_DS, SUBJS
+
+# %%
+# settings
+from utils import calculate_dPrime, get_behavioral_data
+
+# %%
+# Filepaths and settings
+rejectionThreshold = 0.6  # dprime after which we exclude participants
+fpath_ds = FPATH_DS
+overwrite = False
+fpath_der = fpath_ds / "derivatives"
+
+# %%
+
+subex = []
+dPrimeEx = []
+
+subexBino = []
+binoEx = []
+# based on d'
+for sub in SUBJS:
+    fpath_set = fpath_ds / "sourcedata" / "events" / f"EMP{sub:02}_events.csv"
+    behavior_dat = get_behavioral_data(fpath_set)
+    behavior_dat = behavior_dat.assign(
+        right_wrong=[
+            1 if behav == "correctreject" or behav == "hit" else 0
+            for behav in behavior_dat["behavior"]
+        ]
+    )
+
+    dPrime = calculate_dPrime(behavior_dat)  # get dprime
+    p_val = binom_test(
+        sum(behavior_dat["right_wrong"]),
+        len(behavior_dat["right_wrong"]),
+        p=0.5,
+        alternative="greater",
+    )  # get binmoial test
+
+    # collect subjects to exclude based on d'
+    if dPrime < rejectionThreshold:
+        subex.append(sub)
+        dPrimeEx.append(dPrime)
+
+    # collect subjects to exclude based on binomial test against chance
+    if p_val > 0.001:
+        subexBino.append(sub)
+        binoEx.append(p_val)
+# write excluded ppts with their dprimes int df
+exDat = {"subjectNumber": subex, "dPrime": dPrimeEx}
+exdf = pd.DataFrame(data=exDat)
+exdf.to_json(fpath_der / "bad_subs_dPrime.json")
```

Comment:
why save this as JSON? What comes out this way looks a bit odd, because the dataframe index is written as well:

```json
{"subjectNumber":{"0":14,"1":16,"2":18,"3":31},"dPrime":{"0":0.2117642831,"1":0.7893855229,"2":0.7975497007,"3":0.188012621}}
```

This data could just be:

```yaml
{
    "14": 0.2117642831,
    "16": 0.7893855229,
    ... and so on
}
```

Or it could be a proper tabular data format (TSV file, `sep="\t"`)

I think I would write a single TSV file as the output of this script -- with EACH subj as a row, and then columns being their dprime, their pval in the binom test, and two additional binary columns: `exclude_based_on_dprime` (True/False) and `exclude_based_on_binom` (True/False)

for TSV writing, take care that you don't write the pandas index --> that's ugly and not needed. See:  https://github.com/sappelhoff/ecomp_analysis/blob/2aae27233277fb83d4f23fa0218d5d48ba90a132/beh_modeling.py#L636

Reply:
one more comment on JSON writing -- look at this:

https://github.com/sappelhoff/ecomp_analysis/blob/2aae27233277fb83d4f23fa0218d5d48ba90a132/01_find_bads.py#L104-L110

I usually sort the dict, and pass some parameters that make the file nicely human-readable.

Sorting is important if we start tracking the stuff with version control, we don't want a new version if the content is identical and just the ordering changed.

but as said above, I'd prefer a TSV format in this case